--- a/arch/arm64/boot/dts/mediatek/mt7988a-bananapi-bpi-r4.dtsi
+++ b/arch/arm64/boot/dts/mediatek/mt7988a-bananapi-bpi-r4.dtsi
@@ -36,16 +36,8 @@
 		reg = <0x00 0x40000000 0x00 0x10000000>;
 	};
 
-	/* SFP1 cage (WAN) */
-	sfp1: sfp1 {
-		compatible = "sff,sfp";
-		i2c-bus = <&i2c_sfp1>;
-		los-gpios = <&pio 54 GPIO_ACTIVE_HIGH>;
-		mod-def0-gpios = <&pio 82 GPIO_ACTIVE_LOW>;
-		tx-disable-gpios = <&pio 70 GPIO_ACTIVE_HIGH>;
-		tx-fault-gpios = <&pio 69 GPIO_ACTIVE_HIGH>;
-		rate-select0-gpios = <&pio 21 GPIO_ACTIVE_LOW>;
-		maximum-power-milliwatt = <3000>;
+	trng {
+		compatible = "mediatek,mt7988-rng";
 	};
 
 	gpio-keys {
@@ -86,9 +78,8 @@
 };
 
 &gmac2 {
-	sfp = <&sfp1>;
 	managed = "in-band-status";
-	phy-mode = "usxgmii";
+	phy-mode = "10gbase-r";
 	status = "okay";
 	openwrt,netdev-name = "sfp-wan";
 };
@@ -216,6 +207,12 @@
 	};
 };
 
+&i2c1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&i2c1_pins>;
+	status = "okay";
+};
+
 &i2c2 {
 	pinctrl-names = "default";
 	pinctrl-0 = <&i2c2_1_pins>;
