diff --git a/dma.h b/dma.h
index ac5ae3ce37..2462c5931f 100644
--- a/dma.h
+++ b/dma.h
@@ -76,7 +76,7 @@
 
 #define Q_READ(_q, _field) ({						\
 	u32 _offset = offsetof(struct mt76_queue_regs, _field);		\
-	u32 _val;							\
+	u32 _val = 0;							\
 	if ((_q)->flags & MT_QFLAG_NPU) {				\
 		struct airoha_npu *npu;					\
 									\
diff --git a/channel.c b/channel.c
index 5bb739cd07..9c422a0914 100644
--- a/channel.c
+++ b/channel.c
@@ -347,7 +347,7 @@ void mt76_put_vif_phy_link(struct mt76_phy *phy, struct ieee80211_vif *vif,
 
 	rcu_assign_pointer(mvif->offchannel_link, NULL);
 	dev->drv->vif_link_remove(phy, vif, &vif->bss_conf, mlink);
-	kfree(mlink);
+	kfree_rcu(mlink, rcu_head);
 
 	vif->bss_conf.link_id = 0;
 }
diff --git a/mt7996/dma.c b/mt7996/dma.c
index 6e595ef8cd..fa4a04d7f3 100644
--- a/mt7996/dma.c
+++ b/mt7996/dma.c
@@ -725,7 +725,7 @@ int mt7996_dma_init(struct mt7996_dev *dev)
 		ret = mt7996_init_tx_queues(&dev->phy, MT_TXQ_ID(0),
 					    MT7996_NPU_TX_RING_SIZE,
 					    MT_TXQ_RING_BASE(0) + hif1_ofs,
-					    NULL);
+					    wed);
 	else
 		ret = mt7996_init_tx_queues(&dev->phy,
 					    MT_TXQ_ID(dev->mphy.band_idx),
diff --git a/mt7996/debugfs.c b/mt7996/debugfs.c
index f60d4161fb..ac830c628d 100644
--- a/mt7996/debugfs.c
+++ b/mt7996/debugfs.c
@@ -111,6 +111,7 @@ static const char * const mt76_wcid_flag_names[] = {
 	FLAG(PS),
 	FLAG(4ADDR),
 	FLAG(HDR_TRANS),
+	FLAG(TDLS_PEER),
 #undef FLAG
 };
 
diff --git a/mt76.h b/mt76.h
index 527bef97e122..07955555f84d 100644
--- a/mt76.h
+++ b/mt76.h
@@ -420,6 +420,7 @@ enum mt76_wcid_flags {
 	MT_WCID_FLAG_PS,
 	MT_WCID_FLAG_4ADDR,
 	MT_WCID_FLAG_HDR_TRANS,
+	MT_WCID_FLAG_TDLS_PEER,
 
 	NUM_MT76_WCID_FLAGS,
 };
diff --git a/mt76_connac_mcu.c b/mt76_connac_mcu.c
index 89bd52ea8bf7..f61d0625ef51 100644
--- a/mt76_connac_mcu.c
+++ b/mt76_connac_mcu.c
@@ -491,6 +491,11 @@ void mt76_connac_mcu_wtbl_hdr_trans_tlv(struct sk_buff *skb,
 		htr->to_ds = true;
 		htr->from_ds = true;
 	}
+
+	if (test_bit(MT_WCID_FLAG_TDLS_PEER, &wcid->flags)) {
+		htr->to_ds = false;
+		htr->from_ds = false;
+	}
 }
 EXPORT_SYMBOL_GPL(mt76_connac_mcu_wtbl_hdr_trans_tlv);
 
diff --git a/mt7915/main.c b/mt7915/main.c
index e1d83052aa6d..51643a48ed15 100644
--- a/mt7915/main.c
+++ b/mt7915/main.c
@@ -760,6 +760,9 @@ int mt7915_mac_sta_add(struct mt76_dev *mdev, struct ieee80211_vif *vif,
 	msta->wcid.phy_idx = ext_phy;
 	msta->jiffies = jiffies;
 
+	if (sta->tdls)
+		set_bit(MT_WCID_FLAG_TDLS_PEER, &msta->wcid.flags);
+
 	ewma_avg_signal_init(&msta->avg_ack_signal);
 
 	mt7915_mac_wtbl_update(dev, idx,
diff --git a/mt7921/main.c b/mt7921/main.c
index 3d74fabe7408..d39cb881d75e 100644
--- a/mt7921/main.c
+++ b/mt7921/main.c
@@ -828,6 +828,9 @@ int mt7921_mac_sta_add(struct mt76_dev *mdev, struct ieee80211_vif *vif,
 	msta->deflink.last_txs = jiffies;
 	msta->deflink.sta = msta;
 
+	if (sta->tdls)
+		set_bit(MT_WCID_FLAG_TDLS_PEER, &msta->deflink.wcid.flags);
+
 	ret = mt76_connac_pm_wake(&dev->mphy, &dev->pm);
 	if (ret)
 		return ret;
diff --git a/mt7925/main.c b/mt7925/main.c
index 73d3722739d0..61330e3c18b2 100644
--- a/mt7925/main.c
+++ b/mt7925/main.c
@@ -887,6 +887,9 @@ static int mt7925_mac_link_sta_add(struct mt76_dev *mdev,
 	mlink->wcid.link_valid = !!link_sta->sta->valid_links;
 	mlink->sta = msta;
 
+	if (link_sta->sta->tdls)
+		set_bit(MT_WCID_FLAG_TDLS_PEER, &mlink->wcid.flags);
+
 	wcid = &mlink->wcid;
 	ewma_signal_init(&wcid->rssi);
 	rcu_assign_pointer(dev->mt76.wcid[wcid->idx], wcid);
diff --git a/mt7925/mcu.c b/mt7925/mcu.c
index 22bad3cba8df..333cacfaca9c 100644
--- a/mt7925/mcu.c
+++ b/mt7925/mcu.c
@@ -1091,6 +1091,11 @@ mt7925_mcu_sta_hdr_trans_tlv(struct sk_buff *skb,
 		hdr_trans->to_ds = true;
 		hdr_trans->from_ds = true;
 	}
+
+	if (test_bit(MT_WCID_FLAG_TDLS_PEER, &wcid->flags)) {
+		hdr_trans->to_ds = false;
+		hdr_trans->from_ds = false;
+	}
 }
 
 int mt7925_mcu_wtbl_update_hdr_trans(struct mt792x_dev *dev,
diff --git a/mt7996/main.c b/mt7996/main.c
index a8a6552d49f6..9b8db6efb5ac 100644
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -1164,6 +1164,9 @@ mt7996_mac_sta_init_link(struct mt7996_dev *dev,
 	msta_link->wcid.link_valid = !!sta->valid_links;
 	msta_link->wcid.def_wcid = &msta->deflink.wcid;
 
+	if (link_sta->sta->tdls)
+		set_bit(MT_WCID_FLAG_TDLS_PEER, &msta_link->wcid.flags);
+
 	ewma_avg_signal_init(&msta_link->avg_ack_signal);
 	ewma_signal_init(&msta_link->wcid.rssi);
 
diff --git a/mt7996/mcu.c b/mt7996/mcu.c
index 16420375112d..6aaf3ed94221 100644
--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -2139,6 +2139,11 @@ mt7996_mcu_sta_hdr_trans_tlv(struct mt7996_dev *dev, struct sk_buff *skb,
 		hdr_trans->from_ds = true;
 		hdr_trans->mesh = true;
 	}
+
+	if (test_bit(MT_WCID_FLAG_TDLS_PEER, &wcid->flags)) {
+		hdr_trans->to_ds = false;
+		hdr_trans->from_ds = false;
+	}
 }
 
 static enum mcu_mmps_mode
