Silent devfreq when oc MT7988 to 2Ghz
---
--- a/drivers/devfreq/devfreq.c
+++ b/drivers/devfreq/devfreq.c
@@ -375,9 +375,7 @@ static int devfreq_set_target(struct devfreq *devfreq, unsigned long new_freq,
 	freqs.new = new_freq;
 	devfreq_notify_transition(devfreq, &freqs, DEVFREQ_POSTCHANGE);
 
-	if (devfreq_update_status(devfreq, new_freq))
-		dev_warn(&devfreq->dev,
-			 "Couldn't update frequency transition information.\n");
+	devfreq_update_status(devfreq, new_freq);
 
 	devfreq->previous_freq = new_freq;
 
