--- a/arch/arm64/boot/dts/mediatek/mt7988a.dtsi
+++ b/arch/arm64/boot/dts/mediatek/mt7988a.dtsi
@@ -47,8 +47,8 @@
 			opp-hz = /bits/ 64 <900000000>;
 			opp-microvolt = <850000>;
 		};
-		opp-1080000000 {
-			opp-hz = /bits/ 64 <1080000000>;
+		opp-1320000000 {
+			opp-hz = /bits/ 64 <1320000000>;
 			opp-microvolt = <900000>;
 		};
 	};
@@ -130,8 +130,8 @@
 				opp-hz = /bits/ 64 <1500000000>;
 				opp-microvolt = <850000>;
 			};
-			opp-1800000000 {
-				opp-hz = /bits/ 64 <1800000000>;
+			opp-2200000000 {
+				opp-hz = /bits/ 64 <2200000000>;
 				opp-microvolt = <900000>;
 			};
 		};
